### PR TITLE
Dev: refine the parser and serializer of corosync configuration file

### PR DESCRIPTION
**Pull request by nicholasyang2022**
* Rewrites the parser with separated lexing -> parsing stages and using SAX-alike approach for parsing.
* Implement querying in a separated class `DomQuery`.
* Implement serialization in a separated class `DomSerializer`.

### CODE REVIEW

File: crmsh/corosync_config_format.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        self.text = text
+
+
+class UnbalancedBraceException(ParserException):
+    def __init__(self, lineno):
+        super().__init__(f'Unbalanced brace at line {lineno}')
+        self.lineno = lineno
+
+
+class DomParser(Parser):
```
Comment (liangxin1300) [inline crmsh/corosync_config_format.py:99]: Please add docstring here

File: crmsh/conf_parser.py [hunk 2]
```diff
@@ -228,49 +64,76 @@ def get(self, path, index=0):
...
-        """
-        value = self.get(path, index)
-        if not value:
-            raise ValueError("Cannot find value on path \"{}:{}\"".format(path, index))
-        real_path = self._real_path(path, index)
-        exec("del self._config_inst.{}".format(real_path))
+        try:
+            self._dom_query.remove(path, index)
+        except (KeyError, IndexError):
+            raise ValueError("Cannot find value on path \"{}:{}\"".format(path, index)) from None
```
Comment (liangxin1300) [inline crmsh/conf_parser.py:82]: Worth to add behave test case to cover this exception

File: crmsh/conf_parser.py [hunk 3]
```diff
@@ -228,49 +64,76 @@ def get(self, path, index=0):
...
+            else:
+                match node[key]:
+                    case dict(_) as next_node:
+                        if index > 0 and path_stack in self._sec_names_with_list:
+                            if index == 1:
+                                new_node = dict()
+                                node[key] = [next_node, new_node]
+                                node = new_node
+                            else:
+                                raise IndexError(f'index out of range: {index}')
```
Comment (liangxin1300) [inline crmsh/conf_parser.py:103]: Worth to add a behave test case to cover this exception

And
```suggestion
                                raise IndexError(f'Index out of range: {index}')
```

File: crmsh/conf_parser.py [hunk 4]
```diff
@@ -228,49 +64,76 @@ def get(self, path, index=0):
...
+                        else:
+                            node = next_node
+                    case list(_) as li:
+                        if index > len(li):
+                            raise IndexError(f'index out of range: {index}')
```
Comment (liangxin1300) [inline crmsh/conf_parser.py:108]: Worth to add a behave test case to cover this exception

File: crmsh/conf_parser.py [hunk 5]
```diff
@@ -228,49 +64,76 @@ def get(self, path, index=0):
...
-
-    @classmethod
-    def verify_config_file(cls, config_file=None):
-        """
-        Class method to verify config file
-        """
-        inst = ConfParser(config_file)
-        inst._verify_config_file()
+        except IndexError:
+            raise ValueError(f'Index {index} out of range at path "{path}"') from None
```
Comment (liangxin1300) [inline crmsh/conf_parser.py:136]: Worth to add a behave test case to cover this exception

File: crmsh/corosync.py [hunk 6]
```diff
@@ -263,25 +263,23 @@ def add_node_config(ip_list: typing.List[str]) -> None:
...
     inst = conf_parser.ConfParser()
-    inst.convert2dict()
     name_list = inst.get_all("nodelist.node.ring0_addr")
     index = name_list.index(addr)
     inst.remove("nodelist.node", index)
-    utils.str2file(inst.convert2string(), conf())
+    inst.save()
 
 
 def get_corosync_value(key):
```
Comment (liangxin1300) [inline crmsh/corosync.py:285]: It's better to implement such interface https://github.com/ClusterLabs/crmsh/blob/73eaf02a9ff853e501c34706bedad61aa825c68c/crmsh/corosync.py#L431

File: crmsh/corosync_config_format.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+        with StringIO(data) as ifile:
+            return cls(ifile)
+
+    def __init__(self, ifile):
+        """Parse data form a file-like object."""
+        self.__section_stack = list()
+        self._parse(ifile)
+
+    def enter_section(self, name: str):
+        logger.info('entering: %s', name)
```
[the remark this answers is not in the document]
Comment (nicholasyang2022) [reply crmsh/corosync_config_format.py:33]: It serves as a callback interface and is overridden in subclass `DomParser`.

File: crmsh/corosync_config_format.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+                    self._ofile.write(': ')
+                    self.on_value(item)
+                    self._ofile.write('\n')
+
+    def on_value(self, value):
+        self._ofile.write(str(value))
+
+    def __write_indent(self, n):
+        for _ in range(n):
+            self._ofile.write(self._indent)
```
[the remark this answers is not in the document]
Comment (nicholasyang2022) [reply crmsh/corosync_config_format.py:271]: Added more unit tests and archived 100% in unit tests.

Exception cases are corresponding to malformed `corosync.conf`, which only appears when the file is corrupted by manually editing or buggy code. Are you suggesting to cover these cases in functional tests?

File: crmsh/conf_parser.py (hunk 3 above)
Comment (nicholasyang2022) [reply crmsh/conf_parser.py]: Using lower case in IndexError is consistent with python standard library.

**Comment by nicholasyang2022** [comment]
Changes compared to previous version of this pull request:

* added more unit tests for `corosync_config_format` to archive 100% coverage
* add completors for `crm corosync get` and `crm corosync set`
* move `ConfParser` to `corosync` module and drop `conf_parser` module

File: crmsh/corosync_config_format.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/corosync_config_format.py:271]: That's good enough, thanks!

**Pull request merged**